wrapper.go

```diff
@@ -0,0 +1,135 @@
+package sfxlambda
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"github.com/aws/aws-lambda-go/lambda"
+	"github.com/aws/aws-lambda-go/lambdacontext"
+	"github.com/signalfx/golib/datapoint"
+	"os"
+	"reflect"
+	"strings"
+	"time"
+)
+
+// HandlerFuncWrapper is the interface that wraps the lambda handler function.
+type HandlerFuncWrapper interface {
+	GetWrappedHandlerFunc() func(context.Context, interface{}) (interface{}, error)
+	SendDatapoint(context.Context, *datapoint.Datapoint)
+}
+
+type handlerFuncWrapper struct {
+	wrappedHandlerFunc func(context.Context, interface{}) (interface{}, error)
+	defaultDimensions  map[string]string
+	notColdStart       bool
+}
+
+// NewHandlerFuncWrapper is the HandlerFuncWrapper factory function.
+func NewHandlerFuncWrapper(handlerFunc interface{}) HandlerFuncWrapper {
+	hfw := handlerFuncWrapper{}
+	hfw.wrappedHandlerFunc = func(ctx context.Context, payload interface{}) (interface{}, error) {
+		var response interface{}
+		var err error
+		start := time.Now()
+		if hfw.defaultDimensions, err = getDefaultDimensions(ctx); err == nil {
+			hfw.sendInvocationsDatapoint()
+			hfw.sendColdStartsDatapoint()
+			var payloadBytes, responseBytes []byte
+			if payloadBytes, err = json.Marshal(payload); err == nil {
+				if responseBytes, err = lambda.NewHandler(handlerFunc).Invoke(ctx, payloadBytes); err == nil {
+					if nonErrorReturnType := getNonErrorReturnType(handlerFunc); nonErrorReturnType != nil {
+						response = reflect.New(nonErrorReturnType).Interface()
+						err = json.Unmarshal(responseBytes, &response)
+					}
+				}
+			}
+		}
+		if err != nil {
+			hfw.sendErrorsDatapoint()
+		}
+		hfw.sendDurationDatapoint(time.Since(start))
+		return response, err
+	}
+	return &hfw
+}
+
+// GetWrappedHandlerFunc returns the wrapped lambda handler function.
+func (hfw *handlerFuncWrapper) GetWrappedHandlerFunc() func(context.Context, interface{}) (interface{}, error) {
+	return hfw.wrappedHandlerFunc
+}
+
+// SendDatapoint sends custom metrics to SignalFx. If ctx is nil the background context is used.
+func (hfw *handlerFuncWrapper) SendDatapoint(ctx context.Context, dp *datapoint.Datapoint) {
+	sendDatapoint(ctx, dp)
+}
+
+func getDefaultDimensions(ctx context.Context) (map[string]string, error) {
+	if lambdaContext, ok := lambdacontext.FromContext(ctx); ok {
+		if strings.TrimSpace(lambdaContext.InvokedFunctionArn) == "" {
+			return nil, fmt.Errorf("lambda function arn cannot be blank")
+		}
+		// Expected function arn format arn:aws:lambda:us-east-1:accountId:function:functionName:$LATEST
```

Comment:
This is clearly not true because the logic below shows a lot of variation to this expected format.